File: src/com/cattailsw/nanidroid/NanidroidService.java
package com.cattailsw.nanidroid;

import java.util.LinkedList;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.util.Log;
import java.io.InputStream;

import com.cattailsw.nanidroid.util.NetworkUtil;
import com.cattailsw.nanidroid.util.NarUtil;
import android.content.Intent;
import android.app.NotificationManager;
import android.app.Notification;
import java.io.File;
import android.app.PendingIntent;
import java.util.List;
import java.io.FileOutputStream;

/**
 * Describe class <code>HeadLineSensorService</code> here.
 *
 * @author <a href="mailto:"></a>
 * @version 1.0
 */
public class NanidroidService extends Service {
    private static final String TAG = "HeadLineSensorService";
    private SScriptRunner runner = null;

    public static final int START_HEADLINE_SENSOR = 9000;
    public static final int START_NAR_DL = 9001;

    @Override
    public IBinder onBind(Intent arg0) {
	return null;
    }

    @Override
    public void onCreate() {
	super.onCreate();
    }
    private static final int DEF_TIME = 600000;
    private void startHttpTask(long time) {
	mHandler.sendEmptyMessageDelayed(HTTP_TASK_START, time);
    }

    /**
     * Describe <code>onStartCommand</code> method here.
     *
     * @param intent an <code>Intent</code> value
     * @param n an <code>int</code> value
     * @param n1 an <code>int</code> value
     * @return an <code>int</code> value
     */
    public final int onStartCommand(Intent intent, int flags, int startId){
	handleCommand(intent, startId);
	return START_NOT_STICKY;
    }

    @Override
    public void onStart(Intent intent, int startId) {
	super.onStart(intent, startId);
	handleCommand(intent, startId);
    }

    private void handleCommand(Intent i, int startId){
	String action = i.getAction();

	if ( action == null ) {
	    startHttpTask(1000);
	}
	else if (action.equalsIgnoreCase( Intent.ACTION_RUN ) ) {
	    Uri data = i.getData();
	    if ( data != null ) {
		NarDownloadTask n = new NarDownloadTask(data, startId);
		n.execute(this);		
	    }
	}

    }

    @Override
    public void onDestroy() {
	super.onDestroy();
	if (runner != null)
	    runner.stop();

	Log.d(TAG, "onDestory: called");

	mHandler.removeMessages(HTTP_TASK_START);
    }

    static final private int HTTP_TASK_START = 1;
    private Handler mHandler = new Handler() {
	    @Override
	    public void handleMessage(Message msg) {
		switch (msg.what) {
		case HTTP_TASK_START:
		    new SensingTask().execute(NanidroidService.this);
		    startHttpTask(DEF_TIME);
		    break;
		}
	    }
	};

    private class SensingTask extends AsyncTask<Context, String, String> {

	@Override
	protected void onPreExecute() {
	    // You can't call this in the background thread
	    runner = SScriptRunner.getInstance(NanidroidService.this);
	}

	@Override
	protected String doInBackground(Context... args) {
	    long start = System.currentTimeMillis();
	    LinkedList<String> pageContent;
	    try {

		// Try querying the Bottle log API for today's bottle
		pageContent = SSTPBottleSensor.getPageContent(args[0]);
		Log.d(TAG, "bottle.length() = " + pageContent.size());
		runner.addMsgToQueue(pageContent);
		
		pageContent = BottleLogSensor.getPageContent(args[0]);
		runner.addMsgToQueue(pageContent);

	    } catch (Exception e) {
	    }

	    Log.d(TAG, "time = " + (System.currentTimeMillis() - start)
		  + " [ms]");
	    return "End of conversions";
	}

	public void onPostExecute(String result) {
	    if ( runner != null )
		runner.run();
	}
    }

    private class NarDownloadTask extends AsyncTask<Context, String, String> {
	String targetUrl = null;
	Uri targeturi = null;
	int svcid = -1;
	NarDownloadTask(Uri uri, int sid){
	    targeturi = uri;
	    targetUrl = Uri.decode(targeturi.toString());
	    svcid = sid;
	}

	NarDownloadTask(Uri uri){
	    this(uri, -1);
	}
	
	public String doInBackground(Context... args) {
	    try {
		Context ctx = args[0];
		File extDir = ctx.getExternalCacheDir();
		File targetPath = new File(extDir, targeturi.getLastPathSegment());
		Log.d(TAG, "downloading:"+targetUrl+" to " + targeturi.getLastPathSegment());

		if ( NetworkUtil.exists(args[0], targetUrl) == false ) {
		    Log.d(TAG, "file doesn't exist");
		    return null;
		}

		InputStream is = NetworkUtil.getURLStream(args[0], targetUrl);
		NarUtil.copyFile(is, new FileOutputStream(targetPath));
		
		return targetPath.toString();
	    }
	    catch(Exception e){
		e.printStackTrace();
		return null;
	    }
	}

	public void onPostExecute(String result) {
	    if ( result == null ) {
		Log.d(TAG, "download failed.");
		return;
	    }

	    Log.d(TAG, "download complete?");

	    NotificationManager nm = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

	    Notification n = new Notification(R.drawable.ic_launcher, "Download completed", System.currentTimeMillis());
	    Intent ni = new Intent(NanidroidService.this, Nanidroid.class);
	    ni.setData(Uri.fromFile(new File(result))).putExtra("DL_PKG",0);
	    PendingIntent pi = PendingIntent.getActivity(NanidroidService.this, 0, ni, 0);
	    n.setLatestEventInfo(getApplicationContext(), "Download completed", "dl", pi);
	    n.flags = Notification.FLAG_AUTO_CANCEL;
	    nm.notify(42, n);

	    if ( svcid != -1 )
		stopSelf(svcid);
	}


    }

}

File: src/main/kotlin/com/cattailsw/nanidroid/NanidroidService.kt
package com.cattailsw.nanidroid

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.IBinder
import android.util.Log
import com.cattailsw.nanidroid.durable.DurableNotificationPermissionAcceptance
import com.cattailsw.nanidroid.durable.GhostUpdateWorker
import java.io.File
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

internal fun dispatchGhostUpdateEnqueue(executor: Executor, task: () -> Unit) {
    executor.execute(task)
}

/** Fetches the legacy sensor sources in their established order and isolates source failures. */
internal fun fetchAndQueueSensorMessages(
    fetchSstp: () -> Collection<String>,
    fetchBottleLog: () -> Collection<String>,
    enqueue: (Collection<String>) -> Unit,
): Boolean = try {
    enqueue(fetchSstp())
    enqueue(fetchBottleLog())
    true
} catch (_: Exception) {
    false
}

/** Kotlin owner of foreground downloads, polling, and ghost updates. */
class NanidroidService : Service() {
    private var runner: SScriptRunner? = null
    private val activeForegroundStartIds = HashSet<Int>()
    private val serviceJob = SupervisorJob()
    private val serviceScope = CoroutineScope(serviceJob + Dispatchers.IO)
    private var sensingJob: Job? = null
    private val ghostUpdateEnqueueExecutor = Executors.newSingleThreadExecutor { task ->
        Thread(task, "ghost-update-enqueue")
    }

    override fun onBind(intent: Intent?): IBinder? = null

    private fun startHttpTask(initialDelayMillis: Long) {
        if (sensingJob?.isActive == true) return
        runner = SScriptRunner.getInstance(this)
        sensingJob = serviceScope.launch {
            delay(initialDelayMillis)
            while (isActive) {
                val start = System.currentTimeMillis()
                fetchAndQueueSensorMessages(
                    fetchSstp = { SSTPBottleSensor.getPageContent(this@NanidroidService) },
                    fetchBottleLog = { BottleLogSensor.getPageContent(this@NanidroidService) },
                    enqueue = { pageContent -> runner?.addMsgToQueue(pageContent) },
                )
                Log.d(TAG, "time = ${System.currentTimeMillis() - start} [ms]")
                withContext(Dispatchers.Main.immediate) {
                    runner?.run()
                }
                delay(DEF_TIME)
            }
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        ensureForeground()
        handleCommand(intent, startId)
        return START_NOT_STICKY
    }

    @Suppress("DEPRECATION")
    override fun onStart(intent: Intent?, startId: Int) {
        super.onStart(intent, startId)
        handleCommand(intent, startId)
    }

    private fun handleCommand(intent: Intent?, startId: Int) {
        if (intent == null) {
            finishForegroundWork(startId)
            return
        }
        val action = intent.action
        when {
            action == null -> startHttpTask(1_000)
            action.equals(Intent.ACTION_RUN, ignoreCase = true) -> {
                Log.w(TAG, "Archive downloads are handled by NarDownloadRepository")
                finishForegroundWork(startId)
            }
            Intent.ACTION_SYNC.equals(action, ignoreCase = true) -> {
                val homeurl = intent.data
                val gid = intent.getStringExtra(EXT_GID)
                val ghostRoot = intent.getStringExtra(EXT_GROOT)
                if (isHttpsUri(homeurl) && gid != null && ghostRoot != null) {
                    startForegroundWork(startId)
                    dispatchGhostUpdateEnqueue(ghostUpdateEnqueueExecutor) {
                        try {
                            val accepted = GhostUpdateWorker.enqueue(this, homeurl!!, gid, File(ghostRoot))
                            if (accepted) DurableNotificationPermissionAcceptance.markAccepted()
                            if (!accepted) {
                                Log.w(TAG, "Ghost update is already active or could not be queued")
                            }
                        } catch (error: RuntimeException) {
                    Log.e(TAG, "Could not enqueue ghost update", error)
                        } finally {
                            finishForegroundWork(startId)
                        }
                    }
                } else {
                    Log.w(TAG, "Rejected update request without an HTTPS URL, ghost id, and root")
                    finishForegroundWork(startId)
                }
            }
            action.equals(ACTION_CAN_STOP, ignoreCase = true) -> finishForegroundWork(startId)
        }
    }

    private fun startForegroundWork(startId: Int) {
        synchronized(activeForegroundStartIds) { activeForegroundStartIds.add(startId) }
    }

    private fun finishForegroundWork(startId: Int) {
        val noForegroundWorkRemains = synchronized(activeForegroundStartIds) {
            activeForegroundStartIds.remove(startId)
            activeForegroundStartIds.isEmpty()
        }
        if (noForegroundWorkRemains) stopForeground(true)
        stopSelf(startId)
    }

    private fun ensureForeground() {
        if (Build.VERSION.SDK_INT >= 26) createNotificationChannelCompat()
        startForeground(FOREGROUND_NOTIFICATION_ID, createForegroundNotification())
    }

    private fun createForegroundNotification(): Notification {
        val launchIntent = Intent(this, Nanidroid::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        val contentIntent = PendingIntent.getActivity(
            this,
            0,
            launchIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
        val builder = Notification.Builder(this)
            .setSmallIcon(R.drawable.notification)
            .setContentTitle(getString(R.string.download_in_progress))
            .setContentText(getString(R.string.download_in_progress))
            .setContentIntent(contentIntent)
            .setOngoing(true)
        if (Build.VERSION.SDK_INT >= 26) {
            try {
                builder.javaClass.getMethod("setChannelId", String::class.java).invoke(builder, CHANNEL_ID)
            } catch (e: Exception) {
                Log.w(TAG, "notification channel API unavailable", e)
            }
        }
        @Suppress("DEPRECATION")
        return builder.notification
    }

    private fun createNotificationChannelCompat() {
        try {
            val channelClass = Class.forName("android.app.NotificationChannel")
            val channel = channelClass.getConstructor(String::class.java, CharSequence::class.java, Int::class.javaPrimitiveType)
                .newInstance(CHANNEL_ID, getString(R.string.download_channel_name), 2)
            val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            manager.javaClass.getMethod("createNotificationChannel", channelClass).invoke(manager, channel)
        } catch (e: Exception) {
            Log.w(TAG, "notification channel API unavailable", e)
        }
    }

    override fun onDestroy() {
        sensingJob?.cancel()
        serviceJob.cancel()
        ghostUpdateEnqueueExecutor.shutdown()
        runner?.stop()
        Log.d(TAG, "onDestory: called")
        super.onDestroy()
    }

    companion object {
        private const val TAG = "HeadLineSensorService"
        private const val DEF_TIME = 600_000L
        private const val CHANNEL_ID = "nanidroid_downloads"
        private const val FOREGROUND_NOTIFICATION_ID = 41

        private fun isHttpsUri(uri: Uri?): Boolean =
            uri != null && "https".equals(uri.scheme, ignoreCase = true) && !uri.host.isNullOrEmpty()

        const val START_HEADLINE_SENSOR = 9000
        const val START_NAR_DL = 9001
        const val ACTION_CAN_STOP = "canstopsensing"
        const val EXT_GID = "ghost_id_to_update"
        const val EXT_GROOT = "ghost_root_to_update"
        @JvmStatic
        fun createUpdateIntent(
            ctx: Context,
            homeurl: String,
            ghostId: String,
            ghostRoot: String,
        ): Intent =
            Intent(ctx, NanidroidService::class.java).apply {
                action = Intent.ACTION_SYNC
                data = Uri.parse(homeurl)
                putExtra(EXT_GID, ghostId)
                putExtra(EXT_GROOT, ghostRoot)
            }
    }
}
